contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/AbstractHttpHandler.java

```diff
@@ -0,0 +1,235 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Base class for handling request on http client and server.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ */
+abstract class AbstractHttpHandler<Q, P> {
+
+  /** The {@link HttpExtractor} used to extract information from request/response. */
+  @VisibleForTesting final HttpExtractor<Q, P> extractor;
+
+  @VisibleForTesting final StatusCodeConverter statusConverter;
+  private final Meter meter;
+
+  /**
+   * Constructor to allow access from subclasses in the same package only.
+   *
+   * @param extractor the implementation of HTTP extractor which handles the particular classes used
+   *     to hold HTTP request and response info in the library being instrumented.
+   */
+  AbstractHttpHandler(HttpExtractor<Q, P> extractor) {
+    this(
+        extractor,
+        new StatusCodeConverter(),
+        OpenTelemetry.getMeterFactory().get(INSTRUMENTATION_LIB_ID));
+  }
+
+  /**
+   * Constructor to allow access from subclasses in the same package only.
+   *
+   * @param extractor the implementation of HTTP extractor which handles the particular classes used
+   *     to hold HTTP request and response info in the library being instrumented.
+   * @param statusConverter the converter from HTTP status codes to OpenTelemetry statuses or {@code
+   *     null} to use the default.
+   * @param meter the named OpenTelemetry meter to use or {@code null} to use the library default of
+   *     {@code io.opentelemetry.contrib.http}.
+   */
+  AbstractHttpHandler(
+      HttpExtractor<Q, P> extractor,
+      @Nullable StatusCodeConverter statusConverter,
+      @Nullable Meter meter) {
+    checkNotNull(extractor, "extractor is required");
+    this.extractor = extractor;
+    if (statusConverter == null) {
+      this.statusConverter = new StatusCodeConverter();
+    } else {
+      this.statusConverter = statusConverter;
+    }
+    if (meter == null) {
+      this.meter = OpenTelemetry.getMeterFactory().get(INSTRUMENTATION_LIB_ID);
+    } else {
+      this.meter = meter;
+    }
+  }
+
+  /**
+   * A convenience to record a {@link Event} with given parameters.
+   *
+   * @param span the span which this {@code Event} will be added to.
+   * @param type the type of event.
+   * @param uncompressedMessageSize size of the message before compressed or zero if unknown.
+   * @param compressedMessageSize size of the message after compressed or zero if unknown.
+   */
+  static void recordMessageEvent(
+      Span span,
+      long id,
+      AttributeValue type,
+      long uncompressedMessageSize,
+      long compressedMessageSize) {
+    Map<String, AttributeValue> attributes = new HashMap<>();
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_TYPE, type);
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_ID, AttributeValue.longAttributeValue(id));
+    if (compressedMessageSize > 0L) {
```

Comment:
Can you add some javadoc that explains what this method is supposed to be used for, from a semantic perspective, not just a functional one? I mean, the existing doc explains *what* it does, but not why you would want to do this.

Reply:
Done